toucan_connectors/s3/s3_connector.py

```diff
@@ -0,0 +1,164 @@
+    external_id: str = Field(
+        ...,
+        description='This is the external ID you need to use on your AWS policy configuration',
+        ui={'readonly': True},
+    )
+
+    def __init__(self, **data):
+        data['external_id'] = data.get('workspace_id', data.get('external_id'))
+        super().__init__(**data)
+
+    class Config:
+        extra = 'allow'
+        validate_on_assignment = True
+
+    @validator('external_id', pre=True, always=True)
+    def _validate_external_id(cls, value: str, values: dict) -> str:
+        return values.get(
+            'workspace_id', value
+        )  # once set, external_id cannot be changed from the workspace id
+
+    def get_status(self) -> ConnectorStatus:
+        try:
+            self._get_assumed_sts_role()
+        except Exception as sts_exc:
+            return ConnectorStatus(
+                status=False,
+                error=f'Cannot verify connection to S3 and/or AssumeRole failed : {sts_exc}',
+            )
+        return ConnectorStatus(status=True)
+
+    def _forge_url(self, access_key: str, access_secret: str, session_token: str, file: str) -> str:
+        # we encode the strings that may contain special characters
+        access_key = urllib.parse.quote(access_key, safe='')
+        access_secret = urllib.parse.quote(access_secret, safe='')
+        session_token = urllib.parse.quote(session_token, safe='')
+
+        # we control slashes ourselves
+        prefix = (self.prefix or '').lstrip('/')
+        file = file.lstrip('/')
+
+        absolute_path = str(
+            ('/' / Path(self.bucket_name) / Path(prefix or '') / file).absolute()
+        ).removeprefix('/')
+
+        forged_url = f's3://{access_key}:{access_secret}@{absolute_path}'
+
+        return forged_url
+
+    def _retrieve_data(
+        self,
+        data_source: S3DataSource,
+        offset: int = 0,
+        limit: int | None = None,
+    ) -> pd.DataFrame | TextFileReader:
+        credentials = self._get_assumed_sts_role()['Credentials']
+
+        # See the documentation here for the schema :
+        # https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/sts/client/assume_role.html
+        access_key = credentials['AccessKeyId']
+        access_secret = credentials['SecretAccessKey']
+        session_token = credentials['SessionToken']
+
+        reader_kwargs = {
+            **data_source.reader_kwargs,
+            'preview_nrows': limit,
+            'preview_offset': offset,
+        }
+        fetcher_kwargs = data_source.fetcher_kwargs
+        client_kwargs = fetcher_kwargs.get('client_kwargs', {})
+        fetcher_kwargs = {
+            **fetcher_kwargs,
+            'client_kwargs': {**client_kwargs, 'session_token': session_token},
+        }
+
+        session = boto3.Session(
+            aws_access_key_id=access_key,
+            aws_secret_access_key=access_secret,
+            aws_session_token=session_token,
+        )
```

Comment:
should we not add this in a `@cached_property` as well too ? (with lines 89-95)

Reply:
i thought of caching the whole s3 client, but as i don't think its actually creating any http calls in the background it shouldn't be such a big deal and it keeps the code easy to read
might want to double check it doesn't doesn't have any http call, if it does it makes sense to cache it yeah

Reply:
@Sanix-Darker just verified it:
```python
session = boto3.Session(
    aws_access_key_id=access_key,
    aws_secret_access_key=access_secret,
    aws_session_token=session_token,
)
s3 = session.client('s3')
```
- no http request is being made
-  simple class initialisation and set up 
- longest operations are botocore internals JSON files being loaded
-  both session + client init takes about 90ms on my machine (including debug logging)

No need to make the code more complicated and add caching IMO

Reply:
looks great then ! 👌🏽